heaps/min_heap.py

```diff
@@ -57,19 +72,63 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(log n)
+            Space complexity: O(log n)
         """
-        pass
+        if self.empty():
+            return
+
+        if index != 0:
+            compare_index = (index - 1) // 2
+            if self.store[index].key < self.store[compare_index].key:
+                self.swap(index, compare_index)
+                self.heap_up(compare_index)
+        
+        return
 
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
         """
-        pass
+        if self.empty():
+            return
+
+        left_child_index = index * 2 + 1
+        right_child_index = index * 2 + 2
+
+        if left_child_index < len(self.store):
+            if right_child_index < len(self.store):
+                if self.is_left_index_smaller(left_child_index, right_child_index):
+                # smaller_value = self.find_smaller_key(right_child_index, left_child_index)
+                # if smaller_value == left_child_index:
+                    self.swap(index, left_child_index)
+                    self.heap_down(left_child_index)
+                else:
+                    if self.store[index].key > self.store[right_child_index].key:
+                        self.swap(index, right_child_index)
+                        self.heap_down(right_child_index)
+            else:
+                if self.is_left_index_smaller(left_child_index, index):
+                    self.swap(index, left_child_index)
+                    self.heap_down(left_child_index)
```

Comment:
👀 Notice that each time we decide which child to swap, we swap the child, then keep heaping from that newly swapped location (following the larger value down). We could reduce the repetition somewhat by figuring out which index (if any) will be swapped, then at the end, actually do the swap and heap follow.